File: Arduino/Testing/Comunicacion/camera.js
const OpenMV = require('./openmv.js');

class Camera {
    constructor (width = 320, height = 240) {
        this.width = width;
        this.height = height;
    }
    openMV = new OpenMV(this.width, this.height);
    #queryString = ''; 
    #ox = -1;
    #oy = -1;
    #bx = -1;
    #by = -1;
    #yx = -1;
    #yy = -1;

    call () {
        const result = this.openMV.data(this.#queryString);
        this.save(result);
    }
    add (string = '') {
        let query = [];
        for (let i = 0; i < this.#queryString.length; i += 2) {
            query.push(this.#queryString.slice(i, i + 2));
        }
        if (query.includes(string)) return;
        this.#queryString += string;
    }
    save (result = '') {
        let dividedResults = result.split('-');
        dividedResults.shift();
        dividedResults.forEach((value) => {
            value = value.split(':');
            switch (value[0]) {
                case 'ox': this.#ox = value[1]; break;
                case 'oy': this.#oy = value[1]; break;
                case 'bx': this.#bx = value[1]; break;
                case 'by': this.#by = value[1]; break;
                case 'yx': this.#yx = value[1]; break;
                case 'yy': this.#yy = value[1]; break;
            }
        })
    }

    ox () { return this.#ox }
    oy () { return this.#oy }
    bx () { return this.#bx }
    by () { return this.#by }
    yx () { return this.#yx }
    yy () { return this.#yy }
}

module.exports = Camera;

File: Arduino/main_b/Camera.cpp
#include <Arduino.h>
#include "Camera.h"
#define NUMITEMS(arg) ((unsigned int) (sizeof (arg) / sizeof (arg [0])))

bool ongoing_req = false;
Camera::Camera () {
  this -> baud = 115200;
  this -> timeout = 2;
  this -> width = 320;
  this -> height = 240;
}

void Camera::begin() {
  Serial1.begin(baud);
  Serial1.setTimeout(timeout);
}

int Camera::direction(int value, int splits) {
  if (value > 0) {
    if (value <= (width / splits)) return 0;
    for (int i = 1; i < splits; i++) {
      if (value > ((width / splits) * i) && value < ((width / splits) * (i + 1))) return i - 1;
    }
    return splits - 1;
  }
  return -1;
}

bool Camera::call() {
  unsigned long mill = millis();
  Serial1.println(queryString);
  String read = "";
  if (Serial1.available()) { // After many unsuccessfull attempts to retrive information, the camera will eventually (~40ms) return the data back
    read = Serial1.readString();
    save(read);
    return true;
  }
  return false; // If no data has been received, return false so the loop will continue
}

void Camera::del() {
  querySize = 0;
  queryString = "";
}

void Camera::add(String string) {
  String query[queryString.length() / 2];
  for (int i = 0; i < queryString.length(); i += 2) {
    query[i / 2] = queryString.substring(i, i + 2);
  }
  if (includes(query, NUMITEMS(query), string)) return;
  querySize++;
  queryString += string;
}

void Camera::save(String result) {
  String arr[querySize];
  String res = result.substring(1, result.length() - 1);
  int amount_read = 0;
  int last_found = 0;
  for (int i = 0; i < res.length(); i ++) {
    if (res[i] == '#') {
      arr[amount_read] = res.substring(last_found, i);
      amount_read++;
      last_found = i + 1;
    }
  }
  arr[querySize - 1] = res.substring(last_found, res.length());
  for (int i = 0; i < querySize; i++) {
    String object = arr[i];
    String objectArr[2];
    for (int i = 0; i < object.length(); i++) {
      if (object[i] == ':') {
        objectArr[0] = object.substring(0, i);
      }
    }
    objectArr[1] = object.substring(3, object.length());
    if (objectArr[0] == "ox") { _ox = objectArr[1] == "-1" ? -1 : objectArr[1].toInt(); }
    else if (objectArr[0] == "oy") { _oy = objectArr[1] == "-1" ? -1 : objectArr[1].toInt(); }
    else if (objectArr[0] == "bx") { _bx = objectArr[1] == "-1" ? -1 : objectArr[1].toInt(); }
    else if (objectArr[0] == "by") { _by = objectArr[1] == "-1" ? -1 : objectArr[1].toInt(); }
    else if (objectArr[0] == "yx") { _yx = objectArr[1] == "-1" ? -1 : objectArr[1].toInt(); }
    else if (objectArr[0] == "yy") { _yy = objectArr[1] == "-1" ? -1 : objectArr[1].toInt(); }
    else if (objectArr[0] == "bi") { _bi = objectArr[1] == "-1" ? -1 : objectArr[1].toInt(); }
    else if (objectArr[0] == "yi") { _yi = objectArr[1] == "-1" ? -1 : objectArr[1].toInt(); }
    else if (objectArr[0] == "bw") { _bw = objectArr[1] == "-1" ? -1 : objectArr[1].toInt(); }
    else if (objectArr[0] == "yw") { _yw = objectArr[1] == "-1" ? -1 : objectArr[1].toInt(); }
  }
  querySize = 0;
  queryString = "";
}

bool Camera::includes(String arr[], int len, String target) {
  for (int i = 0; i < len; i++) if (arr[i] == target) return true;
  return false;
}